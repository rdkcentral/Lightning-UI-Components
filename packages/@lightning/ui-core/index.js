export { default as utils } from './src/utils/index.js';
export * from './src/components/index.js';
export * from './src/globals/index.js';
export * from './src/mixins/index.js';
export * from './src/shaders/index.js';
export * from './src/textures/index.js';
