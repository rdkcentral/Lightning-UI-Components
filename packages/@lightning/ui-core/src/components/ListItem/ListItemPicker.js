import ListItem from './ListItem';
import lng from '@lightningjs/core';
import * as styles from './ListItemPicker.styles.js';
import Marquee from '../Marquee';
import Row from '../Row';
import { Arrow } from '../../textures';
import utils from '../../utils';
import { withExtensions } from '../../mixins/index.js';

class ListItemPicker extends ListItem {
  static get __componentName() {
    return 'ListItemPicker';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      ...super._template(),
      justify: 'center',
      LeftArrow: {
        mountY: 0.5,
        y: h => h / 2
      },
      RightArrow: {
        mountY: 0.5,
        mountX: 1,
        y: h => h / 2
      }
    };
  }

  static get properties() {
    return [...super.properties, 'options', 'selectedOptionIndex'];
  }

  static get tags() {
    return [
      ...super.tags,
      'LeftArrow',
      'RightArrow',
      { name: 'Picker', path: 'Content.TextWrapper.Picker' }
    ];
  }

  _init() {
    this._options = [];
    this._selectedOptionIndex = 0;
  }

  _update() {
    super._update();
    this._updateArrows();
    this._updatePicker();
  }

  _updateArrows() {
    this._updateLeftArrow();
    this._updateRightArrow();
  }

  _updateLeftArrow() {
    this._LeftArrow.texture = {
      ...this._arrowTexture,
      direction: 'left'
    };
    this.applySmooth(this._LeftArrow, {
      color: this.style.arrowColor,
      x: this.style.paddingX
    });
  }

  _updateRightArrow() {
    this._RightArrow.texture = {
      ...this._arrowTexture,
      direction: 'right'
    };
    this.applySmooth(this._RightArrow, {
      color: this.style.arrowColor,
      x: this.w - this.style.paddingX
    });
  }

  _onTextBoxChanged() {
    super._onTextBoxChanged();
    if (this._Picker) {
      this._Picker.x = (this._Picker.w - this._Title.w) / -2;
    }
  }

  _updatePicker() {
    const w = this.w - this._paddingX - this.style.arrowWidth * 2;

    if (!this._Picker) {
      this._TextWrapper.patch({
        Picker: {
          type: Row,
          clipping: true,
          alwaysScroll: true,
          signals: {
            selectedChange: '_onSelectedChange'
          }
        }
      });
    }
    this._Picker.patch({
      visible: !this._collapse,
      h: this.style.descriptionTextStyle.lineHeight,
      w,
      x: (w - this._Title.w) / -2,
      items: this.options.map(option => ({
        type: Marquee,
        h: this.style.descriptionTextStyle.lineHeight,
        w,
        title: {
          ...this.style.descriptionTextStyle,
          textAlign: 'center',
          text: option
        }
      }))
    });
  }

  _onSelectedChange() {
    if (this.selectedIndex === this.options.length - 1) {
      this._RightArrow.alpha = this.style.arrowAlphaValueLimit;
    } else {
      this._RightArrow.alpha = this.style.arrowAlphaValue;
    }
    if (this.selectedIndex === 0) {
      this._LeftArrow.alpha = this.style.arrowAlphaValueLimit;
    } else {
      this._LeftArrow.alpha = this.style.arrowAlphaValue;
    }
    this.fireAncestors('$announce', this.announce);
  }

  get _arrowTexture() {
    return {
      type: Arrow,
      w: this.style.arrowWidth,
      h: this.style.arrowHeight,
      color: lng.StageUtils.getRgbString(
        utils.getValidColor(this.style.arrowColor)
      )
    };
  }

  get _collapse() {
    return this.shouldCollapse && !this._isFocusedMode;
  }

  get selectedOption() {
    return this._Picker.selected;
  }

  get selectedIndex() {
    return this._Picker.selectedIndex;
  }

  _handleLeft() {
    if (typeof this.onLeft === 'function') {
      return this.onLeft(this);
    }
    this._Picker.selectPrevious();
    return true;
  }

  _handleRight() {
    if (typeof this.onRight === 'function') {
      return this.onRight(this);
    }
    this._Picker.selectNext();
    return true;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }
    return this.title + this.options[this.selectedIndex] + ', List Item';
  }
}

export default withExtensions(ListItemPicker);
