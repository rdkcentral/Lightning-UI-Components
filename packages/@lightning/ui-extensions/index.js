export * from './src/FocusRing';
export * from './src/DropShadow';
