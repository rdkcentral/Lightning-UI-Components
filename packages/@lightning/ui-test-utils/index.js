export * from './src/lightning-test-renderer';
export * from './src/lightning-test-utils';
export * from './src/lng-test-env';
