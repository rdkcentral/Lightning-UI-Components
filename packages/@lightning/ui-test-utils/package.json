{
  "name": "@lightning/ui-test-utils",
  "version": "1.1.0",
  "type": "module",
  "exports": "./index.js",
  "files": [
    "src/*"
  ],
  "peerDependencies": {
    "@lightningjs/core": "^2.7.1"
  },
  "devDependencies": {
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jest-webgl-canvas-mock": "^0.2.3"
  },
  "dependencies": {
    "@lightning/ui-core": "workspace:^",
    "@lightningjs/core": "^2.7.1"
  }
}
