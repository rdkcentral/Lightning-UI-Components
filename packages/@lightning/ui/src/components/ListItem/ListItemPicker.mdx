import { Canvas, Story } from '@storybook/addon-docs';
import ListItemPicker from '.';

# ListItemPicker

ListItem component with the ability to let a user pick from a list of options.

## Source

https://github.comcast.com/Lightning/lightning-ui/packages/@lightning/ui/src/components/ListItemPicker/ListItemPicker.js

## Usage

ListItemPicker requires a title and a list of options

```js
import { ListItemPicker } from '@lightning/ui';

class Basic extends lng.Component {
  static _template() {
    return {
      ListItemPicker: {
        type: ListItemPicker,
        title: 'Title',
        description: 'Description'
      }
    };
  }
}
```

<Canvas>
  <Story id="controls-listitempicker--list-item-picker" />
</Canvas>

## API

### Properties

`ListItemPicker` has the same properties as the ones listed in [ListItem](?path=/docs/lightning/ui-listitem--listitem) in addition to a couple listed below:

| name                | type   | default | description                      |
| ------------------- | ------ | ------- | -------------------------------- |
| options             | any[]  | -       | list of selectable options       |
| selectedOptionIndex | number | -       | index of current selected option |

### Style Properties

| name                 | type   | description                                                          |
| -------------------- | ------ | -------------------------------------------------------------------- |
| arrowAlphaValue      | number | alpha value of the arrows if value is not equal to the min/max value |
| arrowAlphaValueLimit | number | alpha value of the arrows when value is equal to the min/max value   |
| arrowWidth           | number | width of the arrows                                                  |
| arrowHeight          | object | height of the arrows                                                 |
