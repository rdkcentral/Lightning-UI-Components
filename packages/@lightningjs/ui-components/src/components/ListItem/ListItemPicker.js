/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import ListItem from './ListItem';
import * as styles from './ListItemPicker.styles.js';
import Marquee from '../Marquee';
import Icon from '../Icon';
import Row from '../Row';

export default class ListItemPicker extends ListItem {
  static get __componentName() {
    return 'ListItemPicker';
  }

  static get __themeStyle() {
    return styles;
  }

  static get properties() {
    return [...super.properties, 'options', 'selectedIndex'];
  }

  static get tags() {
    return [
      ...super.tags,
      'LeftArrow',
      'RightArrow',
      { name: 'Picker', path: 'Content.TextWrapper.Picker' }
    ];
  }

  _onTextBoxChanged() {
    super._onTextBoxChanged();
    this._alignPicker();
  }

  _construct() {
    super._construct();
    this._options = [];
    this._selectedIndex = 0;
  }

  _update() {
    super._update();
    this._updatePicker();
    this._updateArrows();
    this._updateArrowsAlpha();
    this._updateAlignment();
  }

  _updateAlignment() {
    if (this._isFocusedMode) {
      this.patch({ justify: 'center' });
    } else {
      this.patch({ justify: 'left' });
    }
  }

  _updateArrows() {
    if (!this._isFocusedMode) {
      const offAlpha = 0.001;
      if (this._LeftArrow) {
        this._LeftArrow.alpha = offAlpha;
      }
      if (this._RightArrow) {
        this._RightArrow.alpha = offAlpha;
      }
      return;
    }

    const arrowProps = {
      w: this.style.arrowWidth,
      h: this.style.arrowHeight,
      style: {
        color: this.style.arrowColor
      },
      alpha: this.style.arrowAlphaValue
    };
    if (!this._LeftArrow) {
      this.patch({
        LeftArrow: {
          type: Icon,
          mountY: 0.5,
          y: h => h / 2
        }
      });
    }
    this._LeftArrow.patch({
      ...arrowProps,
      icon: this.style.iconLeftSrc
    });
    this._LeftArrow.smooth = {
      x: this.style.paddingX
    };

    if (!this._RightArrow) {
      this.patch({
        RightArrow: {
          type: Icon,
          mountY: 0.5,
          mountX: 1,
          y: h => h / 2
        }
      });
    }
    this._RightArrow.patch({
      ...arrowProps,
      icon: this.style.iconRightSrc
    });

    this._RightArrow.smooth = {
      x: this.w - this.style.paddingX
    };
  }

  _updatePicker() {
    const w = this.w - this._paddingX - this.style.arrowWidth * 2;

    if (!this._Picker) {
      this._TextWrapper.patch({
        Picker: {
          type: Row,
          clipping: true,
          alwaysScroll: true,
          signals: {
            selectedChange: '_updateArrowsAlpha'
          }
        }
      });
    }
    this._Picker.patch({
      visible: !this._collapse,
      h: this.style.descriptionTextStyle.lineHeight,
      w,
      items: this.options.map(option => ({
        type: Marquee,
        h: this.style.descriptionTextStyle.lineHeight,
        w,
        centerAlign: this._isFocusedMode,
        title: {
          ...this.style.descriptionTextStyle,
          text: option
        }
      })),
      // We need to reset the selected index to ensure it does not get reset to zero when patching items.
      selectedIndex: this.selectedIndex
    });
    this._alignPicker();
  }

  _alignPicker() {
    this._Picker.patch({
      mountX: this._isFocusedMode ? 0.5 : 0,
      x: this._isFocusedMode ? this._Title.w / 2 : 0
    });
  }

  _updateArrowsAlpha() {
    if (this._Picker) {
      this._selectedIndex = this._Picker.selectedIndex;
    }
    const alpha = this._isFocusedMode ? this.style.arrowAlphaValue : 0;
    if (this._RightArrow) {
      this._RightArrow.alpha =
        this.selectedIndex === this.options.length - 1 && this._isFocusedMode
          ? this.style.arrowAlphaValueLimit
          : alpha;
    }
    if (this._LeftArrow) {
      this._LeftArrow.alpha =
        this.selectedIndex === 0 && this._isFocusedMode
          ? this.style.arrowAlphaValueLimit
          : alpha;
    }
    this.fireAncestors('$announce', this.announce);
  }

  get _fixedWordWrapWidth() {
    const wordWrapWidthFocused =
      this.w -
      this._paddingX -
      this.style.arrowWidth * 2 -
      2 * this.style.titlePadding;
    const wordWrapWidth = this.w - this._paddingLeft - this._paddingRight;
    return this._isFocusedMode ? wordWrapWidthFocused : wordWrapWidth;
  }

  get _collapse() {
    return this.shouldCollapse && !this._isFocusedMode;
  }

  get selectedOption() {
    return this._Picker.selected;
  }

  _handleLeft() {
    if (typeof this.onLeft === 'function') {
      return this.onLeft(this);
    }
    this._Picker.selectPrevious();
    return true;
  }

  _handleRight() {
    if (typeof this.onRight === 'function') {
      return this.onRight(this);
    }
    this._Picker.selectNext();
    return true;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }
    return this.title + this.options[this.selectedIndex] + ', List Item';
  }
}
