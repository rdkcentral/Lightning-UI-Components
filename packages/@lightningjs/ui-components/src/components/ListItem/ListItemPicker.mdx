import { Meta, Title } from '@storybook/blocks';
import * as ListItemPickerStories from './ListItemPicker.stories';

<Meta of={ListItemPickerStories} />

<Title />

ListItem component with the ability to let a user pick from a list of options.

## Source

https://github.com/rdkcentral/Lightning-UI-Components/blob/develop/packages/%40lightningjs/ui-components/src/components/ListItem/ListItemPicker.js

## Usage

ListItemPicker requires a title and a list of options

```js
import { ListItemPicker } from '@lightningjs/ui-components';

class Basic extends lng.Component {
  static _template() {
    return {
      ListItemPicker: {
        type: ListItemPicker,
        title: 'Title',
        description: 'Description'
      }
    };
  }
}
```

## API

### Properties

`ListItemPicker` has the same properties as the ones listed in [ListItem](?path=/docs/controls-listitem--list-item) in addition to a couple listed below:

| name          | type     | default   | description                      |
| ------------- | -------- | --------- | -------------------------------- |
| options       | string[] | undefined | list of selectable options       |
| selectedIndex | number   | undefined | index of current selected option |

### Style Properties

| name                 | type   | description                                                          |
| -------------------- | ------ | -------------------------------------------------------------------- |
| arrowAlphaValue      | number | alpha value of the arrows if value is not equal to the min/max value |
| arrowAlphaValueLimit | number | alpha value of the arrows when value is equal to the min/max value   |
| arrowWidth           | number | width of the arrows                                                  |
| arrowHeight          | object | height of the arrows                                                 |
| arrowColor           | string | color of the arrows                                                  |
| iconLeftSrc          | string | path of the icon to use for the left arrow                           |
| iconRightSrc         | string | path of the icon to use for the right arrow                          |
