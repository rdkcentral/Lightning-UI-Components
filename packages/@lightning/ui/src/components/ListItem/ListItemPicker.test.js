import ListItemPicker from './ListItemPicker.js';
import { makeCreateComponent } from '@lightning/ui-test-utils';

const createListItemPicker = makeCreateComponent(ListItemPicker);

describe('ListItemPicker', () => {
  let component, testRenderer;

  beforeEach(() => {
    [component, testRenderer] = createListItemPicker({
      title: 'ListItemPicker',
      description: 'Description'
    });
  });
  afterEach(() => {
    component = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should update Title', () => {
    component.title = 'Title';
    testRenderer.forceAllUpdates();
    expect(component._Title.content).toEqual('Title');
  });

  it('should update Title', () => {
    component.title = '';
    testRenderer.forceAllUpdates();
    expect(component._Title.content).toEqual('');
  });

  it('should update Arrows', () => {
    testRenderer.forceAllUpdates();
    expect(component._LeftArrow.texture.direction).toEqual('left');
    expect(component._RightArrow.texture.direction).toEqual('right');
  });

  it('should update Picker', () => {
    testRenderer.forceAllUpdates();
    expect(component._Picker).toBeDefined();
  });

  it('should update Picker with list of options', () => {
    component.options = ['Description1'];
    testRenderer.forceAllUpdates();
    expect(component._Picker.items.length).toEqual(component.options.length);
  });

  it.skip('Alpha value of arrow when value is equal to min value', () => {
    component.options = ['Description1', 'Description2', 'Description3'];
    testRenderer.forceAllUpdates();
    component._Picker.selectedIndex = 1;
    expect(component._LeftArrow.alpha).toEqual(1);
    component._handleLeft();
    expect(component._LeftArrow.alpha).toEqual(0.6);
  });

  it('Alpha value of arrow when value is equal to max value', () => {
    component.options = ['Description1', 'Description2', 'Description3'];
    testRenderer.forceAllUpdates();
    component._Picker.selectedIndex = 1;
    expect(component._RightArrow.alpha).toEqual(1);
    component._handleRight();
    expect(component._RightArrow.alpha).toEqual(0.6);
  });

  describe('Check HandleLeft & HandleRight', () => {
    it('should reduce selectedIndex when handleLeft is clicked', () => {
      component.options = ['Description1', 'Description2', 'Description3'];
      testRenderer.forceAllUpdates();
      component._Picker.selectedIndex = 2;
      component._handleLeft();
      expect(component.selectedIndex).toEqual(1);
    });
    it('should increase selectedIndex when handleRight is clicked', () => {
      component.options = ['Description1', 'Description2', 'Description3'];
      testRenderer.forceAllUpdates();
      component._Picker.selectedIndex = 1;
      component._handleRight();
      expect(component.selectedIndex).toEqual(2);
    });
    describe('announcer', () => {
      it('should use the title and description as the default announce string', () => {
        component.options = ['Description1', 'Description2', 'Description3'];
        testRenderer.forceAllUpdates();
        expect(component.announce).toEqual(
          `ListItemPicker${component.options[0]}, List Item`
        );
      });

      it('should append List Item to end of announce context', () => {
        component.options = ['Description1', 'Description2', 'Description3'];
        testRenderer.forceAllUpdates();
        component._handleRight();
        component._handleRight();
        expect(component.announce).toEqual(
          `ListItemPicker${component.options[2]}, List Item`
        );
      });

      it('should prefer the announce prop over the default announce', () => {
        const overrideString = 'override announcer string';

        component._announce = overrideString;
        testRenderer.forceAllUpdates();
        expect(component.announce).toEqual(overrideString);
      });
    });
  });
});
