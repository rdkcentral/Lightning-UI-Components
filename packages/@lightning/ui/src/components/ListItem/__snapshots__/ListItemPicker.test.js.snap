// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemPicker renders 1`] = `
{
  "Component": {
    "Background": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Background",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Content": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "TextWrapper": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Description": {
              "active": false,
              "alpha": 0.001,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Text": {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Text",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Text",
                  ],
                  "texture": {
                    "fontFace": "DMSansRegular",
                    "fontSize": 24,
                    "lineHeight": 32,
                    "maxLines": 1,
                    "maxLinesSuffix": "...",
                    "text": "Description",
                    "textBaseline": "bottom",
                    "textColor": 4294375161,
                    "textOverflow": "ellipsis",
                    "type": "TextTexture",
                    "verticalAlign": "middle",
                  },
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 2,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 0,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Description",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "Description",
              ],
              "type": "TextBox",
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "Picker": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Items": {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Items",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Items",
                  ],
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 0,
                  "zIndex": 0,
                },
              },
              "clipping": true,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 32,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Picker",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "Row",
              "tag": [Function],
              "tags": [
                "Picker",
              ],
              "type": "Row",
              "visible": true,
              "w": 322,
              "x": -161,
              "y": 0,
              "zIndex": 0,
            },
            "Title": {
              "active": false,
              "alpha": 0.001,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Text": {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Text",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Text",
                  ],
                  "texture": {
                    "fontFace": "XfinityBrownBold",
                    "fontSize": 28,
                    "letterSpacing": -0.2,
                    "lineHeight": 40,
                    "maxLines": 1,
                    "maxLinesSuffix": "...",
                    "text": "ListItemPicker",
                    "textBaseline": "bottom",
                    "textColor": 4294375161,
                    "textOverflow": "ellipsis",
                    "type": "TextTexture",
                    "verticalAlign": "middle",
                    "wordWrapWidth": 338,
                  },
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 2,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 0,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Title",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "Title",
              ],
              "type": "TextBox",
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "TextWrapper",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "TextWrapper",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Content",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Content",
      ],
      "visible": true,
      "w": 0,
      "x": 193,
      "y": 48,
      "zIndex": 0,
    },
  },
}
`;
